log4j-api/src/main/java/org/apache/logging/log4j/spi/LoggerRegistry.java

```diff
@@ -87,97 +114,224 @@ public void putIfAbsent(final Map<String, T> innerMap, final String name, final
+    /**
+     * Constructs an instance <b>ignoring</b> the given the map factory.
+     *
+     * @param mapFactory a map factory
+     * @deprecated As of version {@code 2.24.1}, planned to be removed!
+     */
+    @Deprecated
+    public LoggerRegistry(@Nullable final MapFactory<T> mapFactory) {
+        this();
     }
 
     /**
-     * Returns an ExtendedLogger.
-     * @param name The name of the Logger to return.
-     * @return The logger with the specified name.
+     * Returns the logger associated with the given name.
+     * <p>
+     * There can be made no assumptions on the message factory of the returned logger.
+     * Callers are strongly advised to switch to {@link #getLogger(String, MessageFactory)} and <b>provide a message factory parameter!</b>
+     * </p>
+     *
+     * @param name a logger name
+     * @return the logger associated with the name
+     * @deprecated As of version {@code 2.24.1}, planned to be removed!
+     * Use {@link #getLogger(String, MessageFactory)} instead.
      */
+    @Deprecated
     public T getLogger(final String name) {
-        return getOrCreateInnerMap(DEFAULT_FACTORY_KEY).get(name);
+        requireNonNull(name, "name");
+        return getLogger(name, null);
     }
 
     /**
-     * Returns an ExtendedLogger.
-     * @param name The name of the Logger to return.
-     * @param messageFactory The message factory is used only when creating a logger, subsequent use does not change
-     *                       the logger but will log a warning if mismatched.
-     * @return The logger with the specified name.
+     * Returns the logger associated with the given name and message factory.
+     * <p>
+     * In the absence of a message factory, there can be made no assumptions on the message factory of the returned logger.
+     * This lenient behaviour is only kept for backward compatibility.
+     * Callers are strongly advised to <b>provide a message factory parameter to the method!</b>
+     * </p>
+     *
+     * @param name a logger name
+     * @param messageFactory a message factory
+     * @return the logger associated with the given name and message factory
      */
-    public T getLogger(final String name, final MessageFactory messageFactory) {
-        return getOrCreateInnerMap(factoryKey(messageFactory)).get(name);
+    public T getLogger(final String name, @Nullable final MessageFactory messageFactory) {
+        requireNonNull(name, "name");
+        readLock.lock();
+        try {
+            final Map<MessageFactory, T> loggerByMessageFactory = loggerByMessageFactoryByName.get(name);
+            if (loggerByMessageFactory == null) {
+                return null;
+            }
+            if (messageFactory != null) {
+                return loggerByMessageFactory.get(messageFactory);
+            }
+            return !loggerByMessageFactory.isEmpty()
+                    ? loggerByMessageFactory.values().iterator().next()
+                    : null;
```

Comment:
I think the problem here is how to deal with a `null` message factory parameter. IMHO opinion we should use:

```
MessageFactory actualMessageFactory = messageFactory != null 
        ? messageFactory 
        : ParameterizedMessageFactory.INSTANCE;
```

This is due to the following reasoning:

- Before `2.23.0` the default message factory was determined by `log4j-api`,
- In `2.23.0` the code (in `AbstractLogger`) was removed, since it caused a circular initialization problem.
- Since `2.24.0` implementations decide, which message factory they want. If they do nothing, they get `ParameterizedMessageFactory`.

If `actualMessageFactory` is not null, we can use `Optional` to get rid of the `null` checks.

Reply:
Fixed in e54dd2d976c36aac0f1d9e7b5dc557073fe813b9.